presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -120,22 +192,126 @@ class PartitionAndSerializeOperator : public Operator {
+  // based on the current state and various factors.
+  ProcessingState getNextState() const {
+    switch (currentState_) {
+      case ProcessingState::kUnknown:
+        // In an unknown state, the next state is also unknown.
+        return ProcessingState::kUnknown;
+      case ProcessingState::kProcessingInputElements: {
+        if (!replicateNullsAndAny_)
+          // If replicateNullsAndAny_ is false, the processing is complete, and
+          // the next state is the "done" state.
+          return ProcessingState::kDone;
+        else if (thisNullCount_ > 0)
+          // If there are null elements (thisNullCount_ > 0), the next state is
+          // the "processing nulls" state.
+          return ProcessingState::kProcessingNulls;
+        else if (totalNullCount_ > 0)
+          // If there are no null elements in the current batch but null
+          // elements existed in the past (totalNullCount_ > 0), the next state
+          // is the "done" state, as no further replication is required (we
+          // don't need to 'replicateAny' as none of the partitions can possibly
+          // be empty).
+          return ProcessingState::kDone;
+        else if (
+            nonEmptyPartitions_.size() < numPartitions_ && totalNullCount_ == 0)
+          // If the number of non-empty partitions is less than the total number
+          // of partitions (nonEmptyPartitions_.size() < numPartitions_) and
+          // there are no null elements in the current batch (totalNullCount_ ==
+          // 0), the next state is the "processing any" state.
+          return ProcessingState::kProcessingAny;
+        else
+          // If none of the above conditions are met, an unknown state is
+          // reached, resulting in an assertion failure.
+          VELOX_UNREACHABLE(
+              "Unknown state reached, can not determine next state");
+      }
+      case ProcessingState::kProcessingNulls: {
+        // If the next partition ID is less than the total number of partitions,
+        // the next state remains in the "processing nulls" state.
+        if (nextPartitionId_ < numPartitions_)
+          return ProcessingState::kProcessingNulls;
+        // Otherwise, the next state is the "done" state.
+        return ProcessingState::kDone;
+      }
+      case ProcessingState::kProcessingAny:
+        // The next state is always the "done" state, as processing any elements
+        // does not require further replication.
+        return ProcessingState::kDone;
+      default:
+        VELOX_UNREACHABLE("Invalid state");
+    }
+  }
+
+  // Since we can replicate any row, we prefer to replicate the smallest.
+  // The trade-off is low shuffle data vs increased time (since finding the
+  // smallest request additional pass of input vector). Hence, we only capture
+  // the smallest value from the first batch of input.
+  void captureReplicateAnyElement(const FlatVector<StringView>& dataVector) {
```

Comment:
I see this method called for every input, but somewhere I saw you mentioning that we capture 'any' row only from the first batch. Would you clarify?

Reply:
You are right, this was getting called multiple times. Updated.